resilience4j-spring-boot2/src/main/java/io/github/resilience4j/metrics/autoconfigure/CustomMeterRegistryAutoConfiguration.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2019 lespinsideg
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.github.resilience4j.metrics.autoconfigure;
+
+import io.github.resilience4j.bulkhead.autoconfigure.BulkheadMetricsAutoConfiguration;
+import io.github.resilience4j.bulkhead.autoconfigure.ThreadPoolBulkheadMetricsAutoConfiguration;
+import io.github.resilience4j.circuitbreaker.autoconfigure.CircuitBreakerMetricsAutoConfiguration;
+import io.github.resilience4j.ratelimiter.autoconfigure.RateLimiterMetricsAutoConfiguration;
+import io.github.resilience4j.retry.autoconfigure.RetryMetricsAutoConfiguration;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.composite.CompositeMeterRegistry;
+import org.springframework.boot.actuate.autoconfigure.metrics.MetricsAutoConfiguration;
+import org.springframework.boot.autoconfigure.AutoConfigureAfter;
+import org.springframework.boot.autoconfigure.AutoConfigureBefore;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+/**
+ * Meter Registry AutoConfiguration for resilience4j-metrics when MetricsAutoConfiguration is disabled.
+ */
+@Configuration
+@ConditionalOnMissingBean(MeterRegistry.class)
+@AutoConfigureAfter(MetricsAutoConfiguration.class)
+@AutoConfigureBefore({BulkheadMetricsAutoConfiguration.class, ThreadPoolBulkheadMetricsAutoConfiguration.class, CircuitBreakerMetricsAutoConfiguration.class, RateLimiterMetricsAutoConfiguration.class
+, RetryMetricsAutoConfiguration.class})
+public class CustomMeterRegistryAutoConfiguration {
```

Comment:
We don't need this class. If `MetricsAutoConfiguration` is excluded, we don't need to make any metrics. 
Please remove this class.

Reply:
Hi @dlsrb6342 I am little bit confused here as in the previous comment it was suggested to create our own meterRegistry without MetricsAutoConfiguration so I have created this class.
In case we remove this class then @ConditionalOnBean(MeterRegistry.class) does not match even with MetricAutoConfiguration enabled and fails to instantiate **MetricsPublisher class.
I am testing with below code:

```@Bean
    @ConditionalOnBean(MeterRegistry.class)
    @ConditionalOnProperty(value = "resilience4j.circuitbreaker.metrics.legacy.enabled", havingValue = "false", matchIfMissing = true)
    @ConditionalOnMissingBean
    public TaggedCircuitBreakerMetricsPublisher taggedCircuitBreakerMetricsPublisher(MeterRegistry meterRegistry) {
        return new TaggedCircuitBreakerMetricsPublisher(meterRegistry);
    }
```

And my application class does not exclude MetricsAutoConfiguration like below:

```@SpringBootApplication
public class TestApplication {
	public static void main(String[] args) {
		SpringApplication.run(TestApplication.class, args);
	}
}
```

I am getting the same error in this scenario:

```Description:
Parameter 0 of method taggedThreadPoolBulkheadMetricsPublisher in io.github.resilience4j.bulkhead.autoconfigure.ThreadPoolBulkheadMetricsAutoConfiguration required a bean of type 'io.micrometer.core.instrument.MeterRegistry' that could not be found.
```

Reply:
> I think **MetricsPublisher doesn't need MetricsAutoConfiguration but MeterRegistry. We can create our own 'MeterRegistry' without 'MetricsAutoConfiguration'.

Sorry to confuse you. I mean **Users** can create their own not us.
We want to make **MetricsPublisher bean only when MeterRegistry Bean is exists. 

Could you check again your test code? If we have `@ConditionalOnBean(MeterRegistry.class) ` annotation, it should not try to make **MetricsPublisher Bean. And also, error log was occurred in `TaggedThreadPoolBulkheadMetricsPublisher`.  Please check `TaggedThreadPoolBulkheadMetricsPublisher` bean has annotation.

Reply:
Hi @dlsrb6342 , The error is coming for all **MetricsPublisher classes. I have double checked that```TaggedThreadPoolBulkheadMetricsPublisher```class has the annotation as ```@ConditionalOnBean(MeterRegistry.class)```. As I stated above, it is not able to instantiate any of **MetrcisPublisher bean even when ```MetricsAutoConfiguration``` is enabled. 
As per my analysis, this is due to ```MeterRegistry```class is not an autoConfiguration class and as per Spring docs about @ConditionalOnBean: https://docs.spring.io/ConditionalOnBean.html

> The condition can only match the bean definitions that have been processed by the application context so far and, as such, **it is strongly recommended to use this condition on auto-configuration classes only.**

If I use ```@ConditionalOnBean(MetricAutoConfiguration.class)```instead of ```@ConditionalOnBean(MeterRegistry.class)```then **MetricsPublisher bean gets initialised when ```MetricAutoConfiguration``` is enabled. When ```MetricAutoConfiguration```is disabled, it doesn't initialise **MetricsPublisher bean without any error which is the behaviour we expect. This is because ```MetricAutoConfiguration``` class is an autoConfiguration class and works fine with ```@ConditionOnBean```annotation.

Let me know your views on this.
Thanks.

Reply:
About spring docs, it means, `@ConditionalOnBaen` annotation should be used in auto-configuration class. It does not mean `@ConditionalOnBean`'s parameter should be auto-configuration class. So, it doesn't matter to use `@ConditionalOnBean(MeterRegistry.class)`.

With your branch. I deleted your `CustomMeterRegistryAutoConfiguration` and published in maven local for test.
And I tested with https://github.com/resilience4j/resilience4j-spring-boot2-demo, it doesn't make any error both `MetricsAutoConfiguration` is enabled and disabled.

Could you add your test code in zip file or on github?

Reply:
Hi @dlsrb6342 , I have removed```CustomMeterRegistryAutoConfiguration```as suggested. I tested with https://github.com/resilience4j/resilience4j-spring-boot2-demo and it doesn't make any error. However it fails the existing unit test case in```MetricsAutoConfigurationTest```class as none of **MetricsPublisher Bean could get initialised. I was talking about this problem in my previous comment as well but I think I couldn't explain well. With failed unit test case, we can see even though ```MetricAutoConfiguration```is enabled, with```@ConditionalOnBean(MeterRegistry.class)```it is not able to initialise any of **MetricsPublisher Bean.
I tried a lot but looks like I am kind of stuck here and need help if I am missing something.
I have added new test cases for scenario when```MetricsAutoConfigurationTest```is disabled in the PR.

Reply:
Hi @KrnSaurabh
I tested with your branch. And there is weird log for `MetricsAutoConfigurationTest`.
```console
Exclusions:
-----------

    org.springframework.boot.autoconfigure.web.servlet.error.ErrorMvcAutoConfiguration

    org.springframework.boot.actuate.autoconfigure.metrics.MetricsAutoConfiguration
```

I think `TestApplication` scanned `TestApplicationWithoutMetricsAutoConfiguration`. So `MetricsAutoConfiguration` is excluded.

Please remove `TestApplicationWithoutMetricsAutoConfiguration` and use `TestApplication` in `MeterRegistryWithoutMetricsAutoConfigurationTest` like below.
```java
@RunWith(SpringJUnit4ClassRunner.class)
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT, classes = TestApplication.class)
@EnableAutoConfiguration(exclude = {MetricsAutoConfiguration.class, ErrorMvcAutoConfiguration.class})
public class MeterRegistryWithoutMetricsAutoConfigurationTest {
    ...
}
```

And also, we have to add `SimpleMetricsExportAutoConfiguration.class` to `@AutoConfigureAfter` in `**MetricsAutoConfiguration` because we need `meterRegistry` bean. you can refer [`WebFluxMetricsAutoConfiguration`](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/web/reactive/WebFluxMetricsAutoConfiguration.java).

Reply:
Hi @dlsrb6342 , Apologies for the delayed response, I was occupied with some project work lately. So I tested ```MetricsAutoConfigurationTest``` as suggested above. I removed ```TestApplicationWithoutMetricsAutoConfiguration```and used ```TestApplication```. I found all test cases in ```MetricsAutoConfigurationTest``` failing again same way. Then I removed ```MeterRegistryWithoutMetricsAutoConfigurationTest``` tests as well from the project and ran only existing tests but ```MetricsAutoConfigurationTest``` still fails.